cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,916 @@
+                                 });
+
+      num_invalid_edges = static_cast<size_t>(
+        thrust::distance(invalid_transposed_edge_triangle_count_first,
+                         transposed_edge_triangle_count_pair_first + edgelist_srcs.size()));
+
+      if (num_invalid_edges == 0) { break; }
+      auto num_valid_edges = edgelist_srcs.size() - num_invalid_edges;
+
+      // case 1. For the (p, q), find intersection 'r' to create (p, r, -1) and (q, r, -1)
+      // FIXME: check if 'invalid_transposed_edge_triangle_count_first' is necessery as I operate on
+      // 'vertex_pair_buffer' which contains the ordering with the number of triangles.
+      // FIXME: debug this stage. There are edges that have been removed that are still found in nbr
+      // intersection
+
+      // nbr_intersection requires the edges to be sort by 'src'
+      // sort the invalid edges by src for nbr intersection
+      thrust::sort_by_key(handle.get_thrust_policy(),
+                          edge_first + num_valid_edges,
+                          edge_first + edgelist_srcs.size(),
+                          num_triangles.begin() + num_valid_edges);
+      
+      auto [intersection_offsets, intersection_indices] = detail::nbr_intersection(
+        handle,
+        cur_graph_view,
+        cugraph::edge_dummy_property_t{}.view(),
+        thrust::make_zip_iterator(edgelist_srcs.begin() + num_valid_edges,
+                                  edgelist_dsts.begin() + num_valid_edges),
+        thrust::make_zip_iterator(edgelist_srcs.end(), edgelist_dsts.end()),
+        std::array<bool, 2>{true, true},
+        do_expensive_check);
+
+      rmm::device_uvector<edge_t> num_triangles_invalid(num_invalid_edges, handle.get_stream());
+
+      // Update the number of triangles of each (p, q) edges by looking at their intersection
+      // size.
+      thrust::for_each(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator<edge_t>(0),
+        thrust::make_counting_iterator<edge_t>(num_invalid_edges),
+        [num_triangles = raft::device_span<edge_t>(num_triangles.data() + num_valid_edges, num_invalid_edges), intersection_offsets = raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size())]__device__(auto i) {
+    
+          num_triangles[i] -= intersection_offsets[i + 1] - intersection_offsets[i];
+        });
+
+      auto pair_ = thrust::make_tuple(6, 1);
+      auto itr_ = thrust::find(edge_first + num_valid_edges, edge_first + edgelist_srcs.size(), pair_);
+      
+      auto dist_ = thrust::distance(edge_first + num_valid_edges, itr_);
+
+      // FIXME: Find a way to not have to maintain a dataframe_buffer
+      auto vertex_pair_buffer_p_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(intersection_indices.size(),
+                                                                     handle.get_stream());
```

Comment:
You can double the buffer size, and you can combine the two tabulate calls below to one and the two for_each calls to one as well.

Reply:
```
auto p_r_and_q_r_edge_buffer = allocate_dataframe_buffer<...>(intersection_indices.size() * 2, handle.get_stream());

auto p_r_and_q_r_edge_pair_first = thrust::make_zip_iterator(p_r_and_q_r_edge_buffer.begin(), p_r_and_q_r_edge_buffer.begin() + intersection_indices.szie());
thrust::tabulate(
  handle.get_thrust_policy(),
  p_r_and_q_r_edge_pair_first,
  p_r_and_q_r_edge_pair_first + intersection_indices.size(),
  generate_pr_and_q_r_from_p_q<...>{...});
```